[Merged by Bors] - ci: Add documentation/status page for workflows (#35112)
leanprover-community/mathlib4: docs/workflows.md

# GitHub Workflows Summary

### Note

- The grouping below is done mainly by primary trigger (the main event that causes runs).
- _Importance_ is a rough evaluation of the importance of the workflow in our operations:
  - High: core CI logic, merge-flow, branch-health, or high-impact automation.
  - Medium: quality checks, reports, notifications, and operational helpers.
  - Low: cosmetic/status/bookkeeping helpers with limited blast radius.

### Trigger type glossary

> - `pull_request`: runs on PR activity using the PR branch workflow file.

Reviewer: This is an (intentionally) incomplete listing, so let's make that explicit.

```suggestion
This is an overview of the most important triggers used by Mathlib CI. For a complete listing, please [see the GitHub actions documentation](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows).

- `pull_request`: runs on PR activity using the PR branch workflow file.
```

- `pull_request_target`: runs on PR activity using the base branch workflow file (safer access to repository secrets/write operations when guarded correctly).

> - `merge_group`: runs for GitHub merge queue batches.

Reviewer: > Looks like this is present in some of the triggers (actionlint.yml, build.yml somehow). Should we remove it?

I would be in favour (to do so in a different PR), since this is presumably a leftover from experiments, and we have no idea how it would actually work.

- `issue_comment`: runs when issue or PR comments are created/edited (depending on `types`).
- `pull_request_review`: runs on PR review submissions.
- `pull_request_review_comment`: runs on inline PR review comments.
- `push`: runs when commits are pushed to matching branches/tags.
- `schedule`: runs on a cron schedule in UTC.
- `workflow_dispatch`: manual run from the Actions UI or API.

> - `workflow_run`: runs after another workflow run reaches specified activity types (for example `completed`).

Reviewer: ```suggestion
- `workflow_run`: runs after another workflow run is requested or is completed.
```

| [`lint_and_suggest_pr.yml`](../.github/workflows/lint_and_suggest_pr.yml) | lint and suggest<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/lint_and_suggest_pr.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/lint_and_suggest_pr.yml) | Low | `pull_request` | Runs style linting in suggest mode on pull requests. |
| [`PR_summary.yml`](../.github/workflows/PR_summary.yml) | Post PR summary comment<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/PR_summary.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/PR_summary.yml) | Low | `pull_request_target` | On `pull_request_target`, computes PR summary data (imports/declarations/tech debt), manages related labels, and updates PR comments. |

> | [`add_label_from_diff.yaml`](../.github/workflows/add_label_from_diff.yaml) | Autolabel PRs<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/add_label_from_diff.yaml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/add_label_from_diff.yaml) | Low | `pull_request_target, push` | Runs `lake exe autolabel` and applies an inferred topic label to newly opened PRs. |

Reviewer: ```suggestion
| [`add_label_from_diff.yaml`](../.github/workflows/add_label_from_diff.yaml) | Autolabel PRs<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/add_label_from_diff.yaml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/add_label_from_diff.yaml) | Low | `pull_request_target, push` | Applies an inferred topic label to newly opened PRs using `lake exe autolabel`. |
```

| [`labels_from_comment.yml`](../.github/workflows/labels_from_comment.yml) | Label PR based on Comment<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/labels_from_comment.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/labels_from_comment.yml) | Medium | `issue_comment, pull_request_review, pull_request_review_comment` | Adds/removes an allowlisted set of labels based on comment/review text commands. |
| [`bot_fix_style.yaml`](../.github/workflows/bot_fix_style.yaml) | bot fix style<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/bot_fix_style.yaml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/bot_fix_style.yaml) | Low | `issue_comment, pull_request_review, pull_request_review_comment` | Responds to review/comment events and runs `lint-style-action` in `fix` mode. |

## Push-triggered workflows

Primary trigger for this section: repository push events (`push`).

| File | Name | Importance | Triggers | Description |
|---|---|---|---|---|

> | [`zulip_emoji_merge_delegate.yaml`](../.github/workflows/zulip_emoji_merge_delegate.yaml) | Zulip emoji merge update<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/zulip_emoji_merge_delegate.yaml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/zulip_emoji_merge_delegate.yaml) | Low | `push` | On push, detects merged/delegated PR context and updates Zulip emoji state. |

Author: I moved this under "Maintainer commands" to avoid the clutter

| [`update_dependencies_zulip.yml`](../.github/workflows/update_dependencies_zulip.yml) | Monitor Dependency Update Failures<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/update_dependencies_zulip.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/update_dependencies_zulip.yml) | High | `workflow_run` | Watches dependency-update CI runs and sends Zulip success/failure messages with PR/label handling. |
| [`maintainer_bors_wf_run.yml`](../.github/workflows/maintainer_bors_wf_run.yml) | Add "ready-to-merge" and "delegated" label (workflow_run)<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/maintainer_bors_wf_run.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/maintainer_bors_wf_run.yml) | Medium | `workflow_run` | Consumes artifacts from the maintainer-label workflow and applies/removes labels plus Zulip emoji updates. |

> | [`maintainer_merge_wf_run.yml`](../.github/workflows/maintainer_merge_wf_run.yml) | Maintainer merge (workflow_run)<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/maintainer_merge_wf_run.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/maintainer_merge_wf_run.yml) | Medium | `workflow_run` | Workflow-run follow-up that reads maintainer-merge artifacts and posts notifications/comments. |

Reviewer: ```suggestion
| [`maintainer_merge_wf_run.yml`](../.github/workflows/maintainer_merge_wf_run.yml) | Maintainer merge (workflow_run)<br>[![passed/failed](https://img.shields.io/github/actions/workflow/status/leanprover-community/mathlib4/maintainer_merge_wf_run.yml?label=status)](https://github.com/leanprover-community/mathlib4/actions/workflows/maintainer_merge_wf_run.yml) | Medium | `workflow_run` | Manages labels and posts on Zulip for maintainer merge/delegate commands. |
```
